### PR TITLE
61 basesharefield contract calculations should be more generic

### PR DESCRIPTION
**Pull request by aliberkhsl**
#61

### CODE REVIEW

File: contracts/staking/BaseShareField.sol [hunk 1]
```diff
@@ -287,7 +301,8 @@ contract BaseShareField is DAOContract {
 	function interestsPerBlock() public view virtual returns (uint256) {
 		return accAmountPerShare;
 	}
-    function rmul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+
+	function rmul(uint256 x, uint256 y) internal pure returns (uint256 z) {
```
Comment (sirpy) [inline contracts/staking/BaseShareField.sol:305]: indent?

**Review by sirpy — COMMENTED** [review]
duplicate tests for different token precisions
2,8,16

File: contracts/staking/BaseShareField.sol (hunk 1 above)
Comment (aliberkhsl) [reply contracts/staking/BaseShareField.sol:305]: Yeah linting changed it probably

File: contracts/staking/SimpleStaking.sol [hunk 2]
```diff
@@ -66,6 +66,8 @@ contract SimpleStaking is AbstractGoodStaking, StakingToken {
 		setDAO(_ns);
 		token = ERC20(_token);
 		iToken = ERC20(_iToken);
+		decimals = token.decimals(); // Staking token decimals should be same with token's decimals
+		tokenDecimalDifference = 18 - token.decimals();
```
Comment (sirpy) [inline contracts/staking/SimpleStaking.sol:70]: add a test, to make sure this reverts if token.decimals > 18
should be reverted on solidity 8

**Review by sirpy — APPROVED** [review]

File: contracts/staking/SimpleStaking.sol (hunk 2 above)
Comment (aliberkhsl) [reply contracts/staking/SimpleStaking.sol:71]: Changed

File: contracts/staking/BaseShareField.sol [hunk 3]
```diff
@@ -33,7 +33,8 @@ contract BaseShareField is DAOContract {
 	address public shareToken;
 
 	uint256 public lastRewardBlock;
-
+	// Token decimal difference with 18
```
Comment (sirpy) [inline contracts/staking/BaseShareField.sol:36]: not good explanation

File: contracts/staking/SimpleStaking.sol [hunk 4]
```diff
@@ -346,14 +320,10 @@ contract SimpleStaking is AbstractGoodStaking, StakingToken {
...
-		public
-		view
-		returns (uint256)
-	{	
+	function getTokenPriceInUSD(uint256 _amount) public view returns (uint256) {
 		AggregatorV3Interface tokenPriceOracle =
 			AggregatorV3Interface(getTokenUsdOracle());
 		(, int256 tokenPriceinUSD, , , ) = tokenPriceOracle.latestRoundData();
-		return (uint256(tokenPriceinUSD) * _amount) / (10 ** token.decimals()); // tokenPriceinUSD in 8 decimals and _amount is in Token's decimals so we divide it to Token's at the end to reduce 8 decimals back
+		return (uint256(tokenPriceinUSD) * _amount) / (10**token.decimals()); // tokenPriceinUSD in 8 decimals and _amount is in Token's decimals so we divide it to Token's at the end to reduce 8 decimals back
```
Comment (sirpy) [inline contracts/staking/SimpleStaking.sol:327]: keep spaces

**Review by sirpy — APPROVED** [review]

File: contracts/staking/SimpleStaking.sol (hunk 4 above)
Comment (aliberkhsl) [reply contracts/staking/SimpleStaking.sol]: Fixed them

File: contracts/staking/BaseShareField.sol (hunk 3 above)
Comment (aliberkhsl) [reply contracts/staking/BaseShareField.sol]: Changed it

**Pull request merged**